### DISCUSSION TITLE
[AMD/LINUX] Slow generation after boot, but improves drastically by launching a game. (idk what to title this oddity)

### DISCUSSION THREAD

**Discussion by ProtoBelisarius** [vladmandic/sdnext#732] [Q&A]
I have a puzzling problem, that while surmountable, is quite annoying and Im not sure where to find answers to it, so I started here. (GPU:6800XT, Fedora37)
After booting up the PC and starting the webui, generation is really slow, it uses GPU but even at default(512x unipc:10 or eulera) the console writes ~0.55it/s or ~1.5s/it. Hardware monitor shows really low and intermittent usage/activity of the GPU(bright green line  should be straight maxed out at the top like the pinky line):
![generation512](https://user-images.githubusercontent.com/103048431/235925082-edd0797d-21b2-4d89-bf66-ed7de9f066b2.png)

Other UIs work directly after boot, even if slower than what I can get with this version, but Im also using sub quadratic or opt-split there.  ComfyUI and a1111 both get around 4-5it/s(512x eulera:20). (a1111 and ComfyUI have their own venv).

There are a few ways I can gain speed with this UI, I think just loading a model with Blender worked, but that didnt reliably work, what always works is to start a game in via Steam. I havent tested around a lot, but launching and running a game with proton, while the UI is still active(though not generating),  works all the time. 
Once the game ran for a few minutes, I can turn it off and use this UI at 8-9it/s at default(512x eulera:20). 

Once its running fast, nothing else influences it, as long as I dont restart the PC. I thought/think it might be a driver issue, but as neither a1111 or ComfyUI have any problems directly after boot, I think it might be more of a problem with this UI. 

Im willing to thoroughly test this more, but I dont know where to start, so if anyone has an idea or the same Issue Ill try it out.

**Comment by vladmandic** [comment]
that is really curious, but lets step a bit?

Since you're talking about GPU utilization and AMD GPU, I assume this uses ROCm and since you have A1111 and this repo both installed on the same system, its the same ROCm? Can you confirm version?

second, since each repo is running in its own VENV, they probably use private installs of Torch itself, so lets check torch versions for each?
for this repo, you'll see it in console log on startup, but you can also go to system info tab (on a1111 you'd need to install system info extension) to get torch and detected gpu information.

what are the command line flags you're using for each repo?
and since some command line flags here have been moved to ui settings, what is in settings -> cuda?

**Reply by ProtoBelisarius** [reply]
I tried basically every combination of UI and venv, I tried the a1111 venv with your UI and it worked without issue, as it just updates stuff like accelerate/gradio/transformers etc to your desired versions. All three venvs have pytorch 2.0.0 with ROCm 5.4.2.  

Your UI console:
Python 3.10.11 on Linux                                                    
Version: 6f976c35 Tue May 2 21:30:31 2023 -0400                            
Setting environment tuning                                                 
AMD ROCm toolkit detected                                                  
Torch 2.0.0+rocm5.4.2                                                      
Torch backend: AMD ROCm HIP 5.4.22803-474e8620                             
Torch detected GPU: AMD Radeon RX 6800 XT VRAM 16368 Arch (10, 3) Cores 36

a1111 SystemInfo:
Torch: 
-2.0.0+rocm5.4.2 full  nohalf
GPU: 
-device: AMD Radeon RX 6800 XT (1)
-hip: 5.4.22803-474e8620
Libs:
-xformers: unavailable
-accelerate: 0.12.0
-transformers: 4.25.1

a1111 arguments:
python launch.py --precision full --no-half --medvram --opt-split-attention --api --skip-torch-cuda-test --disable-nan-check


Your UI settings cuda:
 "precision": "Autocast",
    "cuda_dtype": "FP16",
    "no_half": false,
    "no_half_vae": false,
    "upcast_sampling": false,
    "disable_nan_check": true,
    "rollback_vae": false,
    "opt_channelslast": false,
    "cudnn_benchmark": false,
    "cuda_allow_tf32": true,
    "cuda_allow_tf16_reduced": true,
    "cuda_compile": false,
    "cuda_compile_mode": "none",

**Comment by vladmandic** [comment]
only differences are that you're starting a1111 with `--no-half` and `--precision full`, but here you don't have the same thing enabled in cuda settings?

**Reply by ProtoBelisarius** [reply]
Yes, but it works? I can generate images without errors or oom, with full precision and no-half I get oom errors when using multidiffusion upscaler. Im mostly working on game textures and getting them up to 4k, with either ultimate sdupscale or multidiffusion upscaler with tiled vae, but full throws oom error. (just tried it out again by turning precision type to full and set a checkmark at no-half, apply, and then killed and restarted the UI).
The only problem I have is that some images have a few kind of black/dark blue/burned in spots in the end result image, but I didnt think that was because of the precision and just due to my settings and/or multidiffusion extension, checkpoint, etc.

**Reply by vladmandic** [reply]
having both disabled is definitely better, i just spotted the differences.

> black/dark blue/burned in spots in the end result image

without knowing more, that sounds like underbaked sampler, try with more steps?

otherwise, if both rocm and torch are identical, this repo sets some cuda tunings on startup, perhaps that interferes? i'm really shooting in the dark, i'm just trying to think of anything that may be different:

```python
    os.environ.setdefault('FORCE_CUDA', '1')
    os.environ.setdefault('ATTN_PRECISION', 'fp16')
    os.environ.setdefault('PYTORCH_CUDA_ALLOC_CONF', 'garbage_collection_threshold:0.9,max_split_size_mb:512')
    os.environ.setdefault('CUDA_LAUNCH_BLOCKING', '0')
    os.environ.setdefault('CUDA_CACHE_DISABLE', '0')
    os.environ.setdefault('CUDA_AUTO_BOOST', '1')
    os.environ.setdefault('CUDA_MODULE_LOADING', 'LAZY')
    os.environ.setdefault('CUDA_DEVICE_DEFAULT_PERSISTING_L2_CACHE_PERCENTAGE_LIMIT', '0')
```

**Reply by ProtoBelisarius** [reply]
I tried i2i at 768x768 with full and no-half, and it threw this:
![oomI2I](https://user-images.githubusercontent.com/103048431/235953420-309ba995-34de-403e-bd56-e155e4e29f82.png)

A few minutes after, with no generating, it worked. Really weird.

But also likely not related to startup slowness.

**Comment by iDeNoh** [comment]
I can confirm my 6700xt has dropped in speed dramatically, I went from around 7.6 it/s to anywhere between 6-10s/it, rolling back to commit 5148c5b returns to full speed, though that was just the commit I was on before I updated this morning.

Here are my cuda settings, in addition I'm using the --listen cmd arg. I can grab more details when I go on lunch here in an hour or so.

"precision": "Autocast",
"cuda_dtype": "FP16",
"no_half": false,
"no_half_vae": false,
"upcast_sampling": true,
"disable_nan_check": true,
"rollback_vae": false,
"opt_channelslast": false,
"cudnn_benchmark": false,
"cuda_allow_tf32": true,
"cuda_allow_tf16_reduced": true,
"cuda_compile": false,
"cuda_compile_mode": "none",

**Reply by iDeNoh** [reply]
OK, in what is quite possibly the most stupid fix in all of history, reverting to a previous commit and then restarting webui manually and letting it upgrade to the newest version (that I had already tested) seems to also work, so...I have no idea.

**Reply by ProtoBelisarius** [reply]
I had this problem as well, some commits are extremely slow. I havent tested enough of them to tell which are the cause. 
Commit c128c0770b31ccb0cf8bc788793f0a4b4f1f6244 and 6f976c358f6f2db4d23af0c210f4dbb01f8a24d4(my current) are working well and fast for me. One of the commits after c12 caused a drastic drop when I tried it yesterday. But im not sure if this is related to my problem. My speed didnt change after I changed back to c12 even though it worked yesterday until I did the game load "trick". I think these are different issues.

**Reply by iDeNoh** [reply]
Interesting, I haven't tried the game load trick yet (haven't installed any games in linux) but if it happens again I'll absolutely try.

**Reply by ProtoBelisarius** [reply]
I dont know if any game works or why. Im using a game that takes up 2-4GB VRAM and uses protonGE 7-55(valves compatibility layer for windows titles) to run. This works pretty reliably. As in 6/6 times till today. 
When its slow for you, do you also have really low activity/usage of your GPU?

**Reply by iDeNoh** [reply]
Actually, I'm not certain, I'll have to look into that next time it happens.

**Reply by elen07zz** [reply]
> OK, in what is quite possibly the most stupid fix in all of history, reverting to a previous commit and then restarting webui manually and letting it upgrade to the newest version (that I had already tested) seems to also work, so...I have no idea.

did you go back and get full speed and then you upgraded to the latest and still at this speed?
isn't that too weird?

In my case, I did checkout 6f976c3 but my speeds are still terrible.

**Reply by iDeNoh** [reply]
I had already been getting full speeds on the older version, when I updated earlier I was getting terrible speeds so I reverted to the last version posted yesterday which also had slow speeds, I decided to go back to the version I was on this morning which brought me up to full speed again, then I upgraded to the latest commit again and it's fine lol. I wonder if something failed on the previous upgrade or something

**Reply by vladmandic** [reply]
This is more than perplexing.
Given that I don't have an AMD GPU available, I defred to community experts on advice how to setup and auto-tune environment - and I'll have to do the same here...
Please post any updates and I'll monitor this thread.

**Reply by iDeNoh** [reply]
> This is more than perplexing. Given that I don't have an AMD GPU available, I defred to community experts on advice how to setup and auto-tune environment - and I'll have to do the same here... Please post any updates and I'll monitor this thread.

II think for the sake of clarity and accuracy we should just disregard my experience until I can recreate the issue and see if it aligns with what is happening with what  @ProtoBelisarius reported

**Comment by vladmandic** [comment]
good to know last confirmed good commit hash, i'll check the differences.

**Comment by Soulreaver90** [comment]
I had experienced a similar oddity when I reinstall the repo from the ground up. Half of the time, my generations are slow at 0.5-1 it/s. The other half of the time, I’m at full speed of 6.5it/s. I have no idea what causes it, no differences at all between the installs, same settings and arguments. I will note, I haven’t tried updating in a week so I’m not sure if this has been fixed or is a related ongoing issue.

**Comment by IllustriousElephant** [comment]
Just chiming in to report the same issue (6700XT).

I did a clean install and rolled back to https://github.com/vladmandic/automatic/commit/5148c5b0ad6a5a9522e57952b428f89d29f64375, but unfortunately that didn't work for me; 512x512 images still took 8 to 10 seconds per step. I also tried playing two games (one via Proton, one natively), but had no success. 

What did work for me was loading up Automatic1111 and generating a few images there; then I was able to generate images at normal speed with this UI.

I then went back to the latest commit, and still have been able to generate images quickly.

Edit: For the avoidance of doubt, I should clarify that I was able to generate images at normal speed with *this* UI after generating with A1111 for a bit.

I was able to replicate this today after a reboot; I was generating at 4.5 s/it on this UI, then generated 24 images with A1111 at normal speed, 6.3 it/s, over a span of about 6 minutes. Then after that, I immediately got normal speeds again on this UI, 7.2 it/s.

**Comment by CapsAdmin** [comment]
I had this same problem but somehow it magically resolved itself. One thingI remeber on linux though is if you use CoreCTRL and set the GPU to power saving mode, the performance will drastically reduce in a similar way. Maybe something is triggering that? 

(I don't have access to my main pc at the moment for a while)

**Reply by vladmandic** [reply]
that's exactly in-line with my thinking - but i don't have amd gpu to play with.

**Reply by iDeNoh** [reply]
I had a bug on windows a few years ago with my older and GPU that kept the GPU clocks limited to their low power state, even if I played games. I wonder if this is a driver level issue. We should see if those affected are seeing a similar behavior with their GPU core/memory clocks

**Comment by ProtoBelisarius** [comment]
So, a lot of these comments seem like answers to a different question.
I opened this discussion because I exclusively have problems with this specific UI directly after boot. A1111 or ComfyUI can run at full speed directly after booting up.

Regardless, I might have been wrong about the game part, in the sense that its not the game, but the time. 
I just tried it out a few times(3) and basically after 45min/1h after starting the UI, I get full speed, doesn't matter what I do on the PC, even if I'm doing nothing and just start the UI. Maybe classic case of correlation=/=causation, because most often I did a few dailies in game and thus spent about 30mins or more there and likely exceeded the time needed for the UI to, idk, wake up.
I will try it out a few more times tomorrow and will report again.
